libraries/GDBStub/src/internal/gdbstub.c

```diff
@@ -111,118 +86,85 @@ struct XTensa_exception_frame_s gdbstub_savedRegs;
-static void ATTR_GDBFN keepWDTalive() {
-	uint64_t *wdtval=(uint64_t*)0x3ff21048;
-	uint64_t *wdtovf=(uint64_t*)0x3ff210cc;
-	int *wdtctl=(int*)0x3ff210c8;
-	*wdtovf=*wdtval+1600000;
-	*wdtctl|=(1<<31);
-}
-
-//Receive a char from the uart. Uses polling and feeds the watchdog.
-static int ATTR_GDBFN gdbRecvChar() {
-	int i;
-	while (((READ_PERI_REG(UART_STATUS(0))>>UART_RXFIFO_CNT_S)&UART_RXFIFO_CNT)==0) {
-		keepWDTalive();
-	}
-	i=READ_PERI_REG(UART_FIFO(0));
-	return i;
-}
+//Stores ps when single-stepping instruction. -1 when not in use.
+static int32_t singleStepPs = -1;
 
-//Send a char to the uart.
-static void ATTR_GDBFN gdbSendChar(char c) {
-	while (((READ_PERI_REG(UART_STATUS(0))>>UART_TXFIFO_CNT_S)&UART_TXFIFO_CNT)>=126) ;
-	WRITE_PERI_REG(UART_FIFO(0), c);
-}
-
-//Send the start of a packet; reset checksum calculation.
-static void ATTR_GDBFN gdbPacketStart() {
-	chsum=0;
-	gdbSendChar('$');
-}
-
-//Send a char as part of a packet
-static void ATTR_GDBFN gdbPacketChar(char c) {
-	if (c=='#' || c=='$' || c=='}' || c=='*') {
-		gdbSendChar('}');
-		gdbSendChar(c^0x20);
-		chsum+=(c^0x20)+'}';
-	} else {
-		gdbSendChar(c);
-		chsum+=c;
-	}
+//Uart libs can reference these to see if gdb is attaching to them
+bool gdbstub_has_putc1_control() {
+#if GDBSTUB_REDIRECT_CONSOLE_OUTPUT
+	return true;
+#else
+	return false;
+#endif
 }
-
-//Send a string as part of a packet
-static void ATTR_GDBFN gdbPacketStr(char *c) {
-	while (*c!=0) {
-		gdbPacketChar(*c);
-		c++;
-	}
+bool gdbstub_has_uart_isr_control() {
+#if GDBSTUB_CTRLC_BREAK
+	return true;
+#else
+	return false;
+#endif
 }
 
-//Send a hex val as part of a packet. 'bits'/4 dictates the number of hex chars sent.
-static void ATTR_GDBFN gdbPacketHex(int val, int bits) {
-	char hexChars[]="0123456789abcdef";
-	int i;
-	for (i=bits; i>0; i-=4) {
-		gdbPacketChar(hexChars[(val>>(i-4))&0xf]);
-	}
+//Small function to feed the hardware watchdog. Needed to stop the ESP from resetting
+//due to a watchdog timeout while reading a command.
+static void ATTR_GDBFN keepWDTalive() {
+	uint64_t *wdtval = (uint64_t*)0x3ff21048;
+	uint64_t *wdtovf = (uint64_t*)0x3ff210cc;
+	int *wdtctl = (int*)0x3ff210c8;
+	*wdtovf = *wdtval + 1600000;
+	*wdtctl |= 1 << 31;
 }
 
-//Finish sending a packet.
-static void ATTR_GDBFN gdbPacketEnd() {
-	gdbSendChar('#');
-	gdbPacketHex(chsum, 8);
-}
 
 //Error states used by the routines that grab stuff from the incoming gdb packet
 #define ST_ENDPACKET -1
 #define ST_ERR -2
 #define ST_OK -3
 #define ST_CONT -4
+#define ST_DETACH -5
 
 //Grab a hex value from the gdb packet. Ptr will get positioned on the end
 //of the hex string, as far as the routine has read into it. Bits/4 indicates
 //the max amount of hex chars it gobbles up. Bits can be -1 to eat up as much
 //hex chars as possible.
-static long ATTR_GDBFN gdbGetHexVal(unsigned char **ptr, int bits) {
+static long /*ATTR_GDBFN*/ gdbGetHexVal(unsigned char **ptr, int bits) {
```

Comment:
Commented ATTR_GDBFN? should it be removed?

Reply:
Same for other functions below.

Reply:
Yes, this is a leftover from me trying things out as I moved code to flash from IRAM.  Removing the commented out parts in my next push.